File: projects/java_linked_list_GUI/List.java
import java.util.Iterator;

class Node<T> implements Comparable<Node<T>> {

    private T data;
    private Node<T> nextNode;

    Node(T data) {

        this.data = data;
        nextNode = null;

    }

    public T get_data() { return data; }

    public void set_data(T newData) { data = newData; }

    public Node<T> get_next_node() { return nextNode; }

    public void set_next_node(Node<T> next) { nextNode = next; }

    @Override
    public int compareTo(Node<T> anotherNode) {

        T toCompare = anotherNode.get_data();

        if (toCompare instanceof String) {

            String dataThis = (String)data;
            String anotherData = (String)toCompare;

            for (int i = 0; i < dataThis.length() && i < anotherData.length(); i++) {

                if ((int)dataThis.charAt(i) == (int)anotherData.charAt(i)) continue;
                else return (int)dataThis.charAt(i) - (int)anotherData.charAt(i);

            }

            if (dataThis.length() < anotherData.length())
                return (dataThis.length() - anotherData.length());

            else if (dataThis.length() > anotherData.length())
                return (dataThis.length() - anotherData.length());

            else return 0;

        }
        else if(toCompare instanceof Integer) {

            return (Integer)data > (Integer)toCompare ? 1 : (Integer)data == (Integer)toCompare ? 0 : -1;
        }
        else if(toCompare instanceof Float) {

            return (Float)data > (Float)toCompare ? 1 : (Float)data == (Float)toCompare ? 0 : -1;
        }
        else throw new Error("Unsupported data type");

    }

}

public class List<T> implements Iterable<T> {

    private int listSize;
    private Node<T> head;
    private Node<T> tail;

    public List() {

        listSize = 0;
        head = null;
        tail = null;

    }

    public Node<T> push_back(T data) {

        Node<T> newNode = new Node<T>(data);

        if (head == null) tail = head = newNode;
        else tail.set_next_node(newNode);

        tail = newNode;
        ++listSize;

        return newNode;
    }

    public void push_ordered(T data) {

        push_back(data);
        sort();

    }

    public T pop_back() {

        if (listSize == 1) {

            T temp = head.get_data();
            remove(0);
            return temp;
        }

        Node<T> iter = head;
        Node<T> popped;

        for (int i = 0; i != (listSize - 2); ++i) iter = iter.get_next_node();

        popped = iter.get_next_node();
        iter.set_next_node(null);
        tail = iter;
        --listSize;

        return popped.get_data();
    }

    public T get_element_data(int index) {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index);

        Node<T> iter = head;

        for (int i = 0; i != index; ++i) iter = iter.get_next_node();

        return iter.get_data();
    }

    public void set_element_data(int index, T data) {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index);

        Node<T> iter = head;

        for (int i = 0; i != index; ++i) iter = iter.get_next_node();

        iter.set_data(data);

    }

    public void insert(int index, T data) {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index);

        Node<T> iter = head;
        Node<T> newNode = new Node<T>(data);

        for (int i = 0; i != index; ++i) iter = iter.get_next_node();

        newNode.set_next_node(iter.get_next_node());
        iter.set_next_node(newNode);
        ++listSize;

    }

    public void remove(int index) {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index);

        Node<T> iter = head;
        Node<T> temp;

        if (index == 0) {

            head = head.get_next_node();
            --listSize;
            return;
        }

        for (int i = 0; i != index - 1; ++i) iter = iter.get_next_node();

        if (index == listSize) {

            tail = iter;
            --listSize;
            return;
        }

        temp = iter.get_next_node();
        iter.set_next_node(temp.get_next_node());
        --listSize;

    }

    private void swap(Node<T> first, Node<T> second) {

        T toCompare = second.get_data();
        second.set_data(first.get_data());
        first.set_data(toCompare);

    }

    public void sort() {

        boolean sorted = false;
        Node<T> first = null;

        while(!sorted) {

            sorted = true;
            first = head;

            for (int i = 0; i < listSize - 1; i++) {

                Node<T> second = first.get_next_node();

                if (first.compareTo(second) >= 1) {

                    swap(first, second);
                    sorted = false;

                }

                first = second;

            }

        }

    }

    public void print() {

        if (head == null) {

            System.out.println("No elements in the list");
            return;
        }

        Node<T> iter = head;

        for (int i = 0; i != listSize; ++i) {

            System.out.println(iter.get_data());
            iter = iter.get_next_node();

        }

    }

    public Node<T> get_head() { return head; }

    public Node<T> get_tail() { return tail; }

    public int get_size() { return listSize; }

    public Iterator<T> iterator() { return new ListIterator<T>(this); }

}

class ListIterator<T> implements Iterator<T> {

    Node<T> currentNode;

    public ListIterator(List<T> list) { currentNode = list.get_head(); }

    public boolean hasNext() { return currentNode != null; }

    public T next() {

        T data = currentNode.get_data();
        currentNode = currentNode.get_next_node();

        return data;
    }

}

File: projects/scala_linked_list/list.scala
package list
import util.control.Breaks._


class Node[T](
    
    private var data: T, 
    private var nextNode: Node[T] = null
    
){

    def get_data(): T = data

    def set_data(data: T): Unit = { this.data = data }

    def get_next_node(): Node[T] = nextNode
    
    def set_next_node(nextNode: Node[T]): Unit = { this.nextNode = nextNode }

}

class List[T](
    
    private var listSize: Int = 0, 
    private var _head: Node[T] = null, 
    private var _tail: Node[T] = null
    
) extends Iterable[T] {

    def push_back(data: T): Node[T] = {

        var newNode = new Node[T](data)

        if (_head == null) {
            
            _head = newNode
            _tail = _head

        }
        else _tail.set_next_node(newNode)

        _tail = newNode
        listSize += 1

        return newNode
    }

    def push_ordered(data: T): Unit = {

        push_back(data)
        merge_sort()
    }

    def pop_back(): T = {

        if (listSize == 1) {

            var temp: T = _head.get_data()
            remove(0)
            return temp
        }

        var iter: Node[T] = _head
        var popped: Node[T] = null
        var i = 0

        while (i != (listSize - 2)) {

            iter = iter.get_next_node()
            i += 1;

        }
        
        popped = iter.get_next_node()
        iter.set_next_node(null)
        _tail = iter
        listSize -= 1

        return popped.get_data()
    }

    def get_element_data(index: Int): T = {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index)

        var iter: Node[T] = _head
        var i = 0

        while (i != index) { 
            
            iter = iter.get_next_node() 
            i += 1

        }

        return iter.get_data()
    }

    def set_element_data(index: Int, data: T): Unit = {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index)

        var iter: Node[T] = _head
        var i = 0

        while (i != index) { 

            iter = iter.get_next_node()
            i += 1

        } 

        iter.set_data(data)

    }

    def insert(index: Int, data: T): Unit = {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index)

        var iter: Node[T] = _head
        var newNode = new Node[T](data)
        var i = 0

        while (i != index) {
            
            iter = iter.get_next_node()
            i += 1

        }

        newNode.set_next_node(iter.get_next_node())
        iter.set_next_node(newNode)
        listSize += 1

    }

    def remove(index: Int): Unit = {

        if (index < 0 || index > listSize) throw new ArrayIndexOutOfBoundsException(index);

        if (index == 0) {

            _head = _head.get_next_node()
            listSize -= 1

        }
        else {

            var i = 0
            var iter: Node[T] = _head

            while (i != index - 1) { 

                iter = iter.get_next_node()
                i += 1

            }

            if (index == listSize) {

                _tail = iter
                listSize -= 1
                
            }
            else {

                var temp: Node[T] = iter.get_next_node()
                iter.set_next_node(temp.get_next_node())
                listSize -= 1

            }
            
        }

    }

    def iterator = new Iterator[T] {

        var currentNode: Node[T] = _head
        var iter = 0

        def hasNext: Boolean = { 

            iter += 1
            return iter != (listSize + 1) 
        }

        def next(): T = {

            var data = currentNode.get_data()
            currentNode = currentNode.get_next_node()

            return data
        }

    }

    private def merge[T](list: List[T], list1: List[T], list2: List[T]): Unit = {

        var leftSize = list1.get_size()
        var rightSize = list2.get_size()

        var i = 0
        var j = 0
        var k = 0

        if (list.get_element_data(0).isInstanceOf[Int]) {

            while (i < leftSize && j < rightSize) {

                if (list1.get_element_data(i).asInstanceOf[Int] <= list2.get_element_data(j).asInstanceOf[Int]) {

                    list.set_element_data(k, list1.get_element_data(i))
                    i += 1

                }
                else {

                    list.set_element_data(k, list2.get_element_data(j))
                    j += 1

                }

                k += 1

            }

        }
        else if (list.get_element_data(0).isInstanceOf[Float]) {

          while (i < leftSize && j < rightSize) {

                if (list1.get_element_data(i).asInstanceOf[Float] <= list2.get_element_data(j).asInstanceOf[Float]) {

                    list.set_element_data(k, list1.get_element_data(i))
                    i += 1

                }
                else {

                    list.set_element_data(k, list2.get_element_data(j))
                    j += 1

                }

                k += 1

            }

        }
        else if (list.get_element_data(0).isInstanceOf[String]) {

            while (i < leftSize && j < rightSize) {

                if (list1.get_element_data(i).asInstanceOf[String] <= list2.get_element_data(j).asInstanceOf[String]) {

                    list.set_element_data(k, list1.get_element_data(i))
                    i += 1

                }
                else {

                    list.set_element_data(k, list2.get_element_data(j))
                    j += 1

                }

                k += 1

            }

        }
        else throw new Error("Unsupported data type")

        while (i < leftSize) {

            list.set_element_data(k, list1.get_element_data(i))
            i += 1
            k += 1

        }

        while (j < rightSize) {

            list.set_element_data(k, list2.get_element_data(j))
            j += 1
            k += 1

        }

    }

    def msort[T](list: List[T]): Unit = {

        var listSize = list.get_size()
  
        if (listSize < 2) return
        else {
              
            var list1 = new List[T]
            var list2 = new List[T]

            var pivot = listSize / 2
    
            for (i <- 0 until pivot) {
            
              list1.push_back(list.get_element_data(i))
            }
    
            for (i <- pivot until listSize) {
            
              list2.push_back(list.get_element_data(i))
            }
    
            msort(list1)
            msort(list2)
    
            merge(list, list1, list2)

        }
        
    }

    def merge_sort(): Unit = {

        msort(this)
    }

    def print(): Unit = {

        if (_head == null) {

            println("No elements in the list")
            return
        }

        var iter: Node[T] = _head
        
        for (i <- 0 until listSize) {

            println(iter.get_data())
            iter = iter.get_next_node()

        }

    }

    def get_head(): Node[T] = { return _head; }

    def get_tail(): Node[T] = { return _tail; }

    def get_size(): Int = { return listSize; }

}